core/domain/exp_jobs_one_off.py

```diff
@@ -490,3 +493,114 @@ def reduce(key, values):
             yield (key, len(values))
         else:
             yield (key, values)
+
+
+class VoiceoverDurationSecondsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-Off Job to set every voiceover's duration to duration_secs with
+    the correct value.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        exploration = exp_fetchers.get_exploration_by_id(item.id)
+        try:
+            exploration.validate()
+        except Exception as e:
+            logging.error(
+                'Exploration %s failed non-strict validation: %s' %
+                (item.id, e))
+            return
+        if (item.states_schema_version <
+                AUDIO_DURATION_SECS_MIN_STATE_SCHEMA_VERSION):
+            yield ('State schema version is not the '
+                   'minimum version expected, 31', 'EXP_ID: %s, '
+                   'STATE_SCHEMA_VERSION: %s' %
+                   (item.id, item.states_schema_version))
+            return
+
+        # Go through each exploration state to find voiceover recordings.
+        # For batching them as commits per explorations.
+        commit_cmds = []
+        exploration_states_changed_count = 0
+        exploration_states_unchanged_count = 0
+        exploration_states_failed_count = 0
+        for state, state_value in item.states.items():
+            states_changed_count = 0
+            voiceovers_mapping = (state_value['recorded_voiceovers']
+                                  ['voiceovers_mapping'])
+            language_codes_to_audio_metadata = voiceovers_mapping.values()
+            for language_codes in language_codes_to_audio_metadata:
+                for audio_metadata in language_codes.values():
+                    # Get files using the filename.
+                    filename = audio_metadata['filename']
+                    if audio_metadata['duration_secs'] == 0.0:
+                        try:
+                            fs = (fs_domain.AbstractFileSystem(
+                                fs_domain.GcsFileSystem(
+                                    AUDIO_ENTITY_TYPE,
+                                    item.id)))
```

Comment:
Optional: You can bring this to the previous line.

Reply:
Ok

Reply:
done